snuba/schemas.py

```diff
@@ -217,42 +183,149 @@ def get_time_series_query_schema_properties(default_granularity: int, default_wi
     'required': ['project'],
+    'additionalProperties': False,
 }
 
-EVENTS_QUERY_SCHEMA['properties'].update({
-    **get_time_series_query_schema_properties(
+
+def get_time_series_extension_properties(default_granularity: int, default_window: timedelta):
+    return {
+        'type': 'object',
+        'properties': {
+            'from_date': {
+                'type': 'string',
+                'format': 'date-time',
+                'default': lambda: (datetime.utcnow().replace(microsecond=0) - default_window).isoformat()
+            },
+            'to_date': {
+                'type': 'string',
+                'format': 'date-time',
+                'default': lambda: datetime.utcnow().replace(microsecond=0).isoformat()
+            },
+            'granularity': {
+                'type': 'number',
+                'default': default_granularity,
+            },
+        },
+        'additionalProperties': False,
+    }
+
+
+@dataclass
+class Request:
+    query: Mapping[str, Any]
+    extensions: Mapping[str, Mapping[str, Any]]
+
+    def __post_init__(self):
+        self.body = ChainMap(self.query, *self.extensions.values())
+
+
+class RequestSchema:
+    def __init__(self, query_schema, extensions_schemas: Mapping[str, Any]):
+        self.__query_schema = query_schema
+        self.__extension_schemas = extensions_schemas
+
+        self.__composite_schema = {
+            'type': 'object',
+            'properties': {},
+            'required': [],
+            'definitions': {},
+            'additionalProperties': False,
+        }
+
+        for schema in itertools.chain([self.__query_schema], self.__extension_schemas.values()):
+            assert schema['type'] == 'object', 'subschema must be object'
+            assert schema['additionalProperties'] is False, 'subschema must not allow additional properties'
+            self.__composite_schema['required'].extend(schema.get('required', []))
+
+            for property_name, property_schema in schema['properties'].items():
+                assert property_name not in self.__composite_schema['properties'], 'subschema cannot redefine property'
+                self.__composite_schema['properties'][property_name] = property_schema
+
+            for definition_name, definition_schema in schema.get('definitions', {}).items():
+                assert definition_name not in self.__composite_schema['definitions'], 'subschema cannot redefine definition'
+                self.__composite_schema['definitions'][definition_name] = definition_schema
+
+        self.__composite_schema['required'] = set(self.__composite_schema['required'])
+
+    def validate(self, value) -> Request:
+        # XXX: Mutates input value!
+        validate(value, self.__composite_schema)
+
+        query = {key: value.pop(key) for key in self.__query_schema['properties'].keys() if key in value}
+
+        extensions = {}
+        for extension_name, extension_schema in self.__extension_schemas.items():
```

Comment:
Here we are building subsets of the original value to validate according to the structure of the subschema. Implicitly additionalProperties would be ignored. Why do we need to set additionalProperties = False in the schema ?

Reply:
Because right now everything is extracted from a single mapping, it's ambiguous which extension would receive the additional properties as part of its data in the `request.extensions` map. I don't think discarding them is the right answer either, and disallowing them entirely is just a simpler way of managing that complexity for now. I think we can make it more lax in the future if we move to a more structured query schema similar to the one discussed above.